### DISCUSSION TITLE
Enable Link-Time Optimization (LTO) for loco-cli

### DISCUSSION THREAD

**Discussion by zamazan4ik** [loco-rs/loco#699] [Ideas]
Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) for `loco-cli`  is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance.

I suggest enabling LTO only for the Release builds to not sacrifice the developers' experience during the working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `release-lto` profile where additionally to regular `release` optimizations LTO also will be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application.

Possibly there are other projects without LTO - didn't check them all. It would be nice to have LTO enabled by default in some way for all Loco tools.

Thank you.

**Comment by jondot** [comment]
Hi, yes it makes sense!
I believe if LTO is off we may have experimented with it and left it off. Would you like to submit a PR to fix it?

**Reply by zamazan4ik** [reply]
> I believe if LTO is off we may have experimented with it and left it off.

It's very interesting. If someone can clarify it here - it would be awesome! Probably, there are some known limitations to enabling LTO.

> Would you like to submit a PR to fix it?

If I will find free time for doing it :)

**Comment by schungx** [comment]
Well, I turned on `LTO=fat` on my small-ish Loco app and it took 3-4 HOURS to link...

I haven't seen a material improvement in speed for the LTO-linked binary...

It is about 1/4 smaller.

**Reply by zamazan4ik** [reply]
Did you try to use `lto = thin` instead?

**Reply by schungx** [reply]
Not really...  does it matter?  I can try...

**Reply by zamazan4ik** [reply]
Yeah, ThinLTO is MUCH faster in practice than Fat LTO
